spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/messaging/TracingMethodMessageHandlerAdapter.java

```diff
@@ -0,0 +1,104 @@
+/*
+ * Copyright 2013-2019 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.cloud.sleuth.instrument.messaging;
+
+import java.util.function.BiConsumer;
+import java.util.function.Consumer;
+
+import brave.Span;
+import brave.Tracer;
+import brave.Tracing;
+import brave.propagation.Propagation;
+import brave.propagation.TraceContext;
+import brave.propagation.TraceContextOrSamplingFlags;
+
+import org.springframework.messaging.Message;
+import org.springframework.messaging.support.MessageHeaderAccessor;
+
+import static brave.Span.Kind.CONSUMER;
+
+/**
+ * Adds tracing extraction to an instance of
+ * {@link org.springframework.messaging.handler.invocation.AbstractMethodMessageHandler} in a
+ * reusable way. When sub-classing a provider specific class of that type you would wrap the
+ * <pre>super.handleMessage(...)</pre> call with a call to this. See
+ * {@link org.springframework.cloud.sleuth.instrument.messaging.SqsQueueMessageHandler} for an
+ * example.
+ *
+ * This implementation also allows for supplying a {@link java.util.function.BiConsumer} instance
+ * that can be used to add queue specific tags and modifications to the span.
+ *
+ * @author Brian Devins-Suresh
+ */
+class TracingMethodMessageHandlerAdapter {
+
+	private Tracing tracing;
+	private Tracer tracer;
+	private TraceContext.Extractor<MessageHeaderAccessor> extractor;
+
+
+	TracingMethodMessageHandlerAdapter(Tracing tracing,
```

Comment:
Having read through more of the messaging code I'm not sure how reusable this would need to be. It seems most other things will be covered by the existing `spring-integration` style instrumentation. I'd love to get input from maybe someone who works on that project (`spring-messaging`/`spring-integration`) on whether maybe the SQS messaging is a special case

Reply:
Looking through the code I can find, this does seem to become useful for `@MessageMapping` and `@SubscribeMapping` which don't look to be covered by the current websockets instrumentation so I think it is useful to leave this here